File: src/lib/ai/state.ts
// src/state.ts
import { BaseMessage } from "@langchain/core/messages";
import { v4 as uuidv4 } from 'uuid';

export interface RemotionAgentState {
  sessionId: string;
  originalPrompt: string;
  enhancedPrompt: string | null;
  
  // RAG related
  ragContext: string | null;
  
  // Code generation and validation
  generatedCode: string | null;
  validationLog: string[]; // For stdout/stderr from validation attempts
  validationStatus: 'pending' | 'success' | 'failure';
  
  // Error handling and retries
  currentErrors: string[]; // Errors from the latest validation attempt
  totalAttempts: number; // Total generation attempts (1 initial + retries)
  
  // Final output
  finalCode: string | null;
  finalUserMessage: string | null; // Message for user if process fails overall
  
  // For conversational context if nodes use LLMs that benefit from it
  // (Not heavily used in this direct generation flow, but good for future)
  messages: BaseMessage[]; 

    animationConfig: AnimationParameters | null; // Add this to the interface

}

export interface AnimationParameters {
    compositionId: string;
    durationInFrames: number;
    fps: number;
    width: number;
    height: number;
  }
 
export function getInitialState(): Omit<RemotionAgentState, 'originalPrompt' | 'sessionId'> {
  return {
    enhancedPrompt: null,
    ragContext: null,
    generatedCode: null,
    validationLog: [],
    validationStatus: 'pending',
    currentErrors: [],
    totalAttempts: 0,
    finalCode: null,
    finalUserMessage: null,
    messages: [],
animationConfig: null,

  };
}

File: python_backend/remotion_agent_python/state.py
from typing import List, Optional, TypedDict, Literal
from langchain_core.messages import BaseMessage
import uuid

class AnimationParameters(TypedDict, total=False):
    compositionId: Optional[str]
    durationInFrames: Optional[int]
    fps: Optional[int]
    width: Optional[int]
    height: Optional[int]

class RemotionAgentState(TypedDict):
    sessionId: str
    originalPrompt: str
    enhancedPrompt: Optional[str]
    
    ragContext: Optional[str] # For RAG results
    
    generatedCode: Optional[str]
    validationLog: List[str]
    validationStatus: Literal['pending', 'success', 'failure']
    
    currentErrors: List[str]
    totalAttempts: int
    
    finalCode: Optional[str]
    finalUserMessage: Optional[str]
    
    messages: List[BaseMessage]
    animationConfig: Optional[AnimationParameters]


def get_initial_state() -> dict: # Omit<RemotionAgentState, 'originalPrompt' | 'sessionId'>
    return {
        "enhancedPrompt": None,
        "ragContext": None,
        "generatedCode": None,
        "validationLog": [],
        "validationStatus": 'pending',
        "currentErrors": [],
        "totalAttempts": 0,
        "finalCode": None,
        "finalUserMessage": None,
        "messages": [],
        "animationConfig": None,
    } 